pkg/apis/resource/validation/validation.go

```diff
@@ -433,179 +507,238 @@ func validateNodeName(name string, fldPath *field.Path) field.ErrorList {
+	} else {
+		allErrs = append(allErrs, validateBasicDevice(*device.Basic, fldPath.Child("basic"))...)
 	}
 	return allErrs
 }
 
-func validateResourceRequests(requests []resource.ResourceRequest, fldPath *field.Path, requestStored bool) field.ErrorList {
+func validateBasicDevice(device resource.BasicDevice, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
-	for i, request := range requests {
-		idxPath := fldPath.Index(i)
-		allErrs = append(allErrs, validateResourceRequestModel(&request.ResourceRequestModel, idxPath, requestStored)...)
-	}
-	if len(requests) == 0 {
-		// We could allow this, it just doesn't make sense: the entire entry would get ignored and thus
-		// should have been left out entirely.
-		allErrs = append(allErrs, field.Required(fldPath, "empty entries with no requests are not allowed"))
+	allErrs = append(allErrs, validateMap(device.Attributes, resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice, validateQualifiedName, validateDeviceAttribute, fldPath.Child("attributes"))...)
+	allErrs = append(allErrs, validateMap(device.Capacity, resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice, validateQualifiedName, validateQuantity, fldPath.Child("capacity"))...)
+	if combinedLen, max := len(device.Attributes)+len(device.Capacity), resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice; combinedLen > max {
+		allErrs = append(allErrs, field.Invalid(fldPath, combinedLen, fmt.Sprintf("the total number of attributes and capacities must not exceed %d", max)))
 	}
 	return allErrs
 }
 
-func validateResourceRequestModel(model *resource.ResourceRequestModel, fldPath *field.Path, requestStored bool) field.ErrorList {
+var (
+	numericIdentifier = `(0|[1-9]\d*)`
+
+	preReleaseIdentifier = `(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)`
+
+	buildIdentifier = `[0-9a-zA-Z-]+`
+
+	semverRe = regexp.MustCompile(`^` +
+
+		// dot-separated version segments (e.g. 1.2.3)
+		numericIdentifier + `\.` + numericIdentifier + `\.` + numericIdentifier +
+
+		// optional dot-separated prerelease segments (e.g. -alpha.PRERELEASE.1)
+		`(-` + preReleaseIdentifier + `(\.` + preReleaseIdentifier + `)*)?` +
+
+		// optional dot-separated build identifier segments (e.g. +build.id.20240305)
+		`(\+` + buildIdentifier + `(\.` + buildIdentifier + `)*)?` +
+
+		`$`)
+)
+
+func validateDeviceAttribute(attribute resource.DeviceAttribute, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
-	entries := sets.New[string]()
-	if model.NamedResources != nil {
-		entries.Insert("namedResources")
-		allErrs = append(allErrs, namedresourcesvalidation.ValidateRequest(namedresourcesvalidation.Options{StoredExpressions: requestStored}, model.NamedResources, fldPath.Child("namedResources"))...)
+	numFields := 0
+	if attribute.BoolValue != nil {
+		numFields++
+	}
+	if attribute.IntValue != nil {
+		numFields++
 	}
-	switch len(entries) {
+	if attribute.StringValue != nil {
+		numFields++
```

Comment:
Check the length here? Also below for semver?

Reply:
Ack.